### PR TITLE
MF 210 : Add edit options to all the results section

### PR DESCRIPTION
**Pull request by donaldkibet**
MF-210

- Added ability to edit vitals

### CODE REVIEW

File: src/widgets/vitals/vitals-card.resource.tsx [hunk 1]
```diff
@@ -6,73 +6,89 @@ import {
...
+    }),
+    map(entries => entries?.map(entry => entry.resource)),
+    map(vitals => {
       return formatVitals(
-        getVitalsByConcept(data, SYSTOLIC_BLOOD_PRESSURE_CONCEPT),
-        getVitalsByConcept(data, DIASTOLIC_BLOOD_PRESSURE_CONCEPT),
-        getVitalsByConcept(data, PULSE_CONCEPT),
-        getVitalsByConcept(data, TEMPERATURE_CONCEPT),
-        getVitalsByConcept(data, OXYGENATION_CONCEPT)
+        vitals?.filter(vital =>
```
Comment (FlorianRappl) [inline src/widgets/vitals/vitals-card.resource.tsx]: I would drop all the `?.` here (and use `.`). Instead I would transform 

```ts
map(entries => entries?.map(entry => entry.resource))
```

to be

```ts
map(entries => entries?.map(entry => entry.resource) ?? [])
```

then we continue with an empty array.

File: src/widgets/vitals/vitals-card.resource.tsx [hunk 2]
```diff
@@ -97,16 +115,18 @@ function formatVitals(
...
+      weight: weight && weight.valueQuantity.value,
+      height: height && height.valueQuantity.value
     });
   });
 }
 
-function getDatesIssued(vitalsArray): string[] {
-  return vitalsArray.map(vitals => vitals.issued);
+function getDatesIssued(vitalsArray): Array<Date> {
+  return vitalsArray?.map(vitals => vitals.issued);
```
Comment (FlorianRappl) [inline src/widgets/vitals/vitals-card.resource.tsx]: If you *always*  return an array (e.g., via `?.map(...) ?? []`) you'd not require all the `?.` belows.

Note: Every `?.` adds to runtime. Its a couple more checks in the transformed code.

Example:

```ts
foo?.bar()
```

is usually transpiled to

```js
foo === null || foo === void 0 ? void 0 : foo.bar();
```

File: src/widgets/vitals/vitals-card.resource.tsx [hunk 3]
```diff
@@ -97,16 +115,18 @@ function formatVitals(
...
 }
 
-function latestFirst(a, b) {
+function latestFirst(a: Date, b: Date) {
   return new Date(b).getTime() - new Date(a).getTime();
```
Comment (FlorianRappl) [inline src/widgets/vitals/vitals-card.resource.tsx:128]: I thought `a` and and `b` are `Date`s already? Shouldn't it just be `b.getTime() - a.getTime()`?

File: src/widgets/vitals/vitals-detailed-summary.component.tsx [hunk 4]
```diff
@@ -95,7 +95,7 @@ export default function VitalsDetailedSummary(
             {currentPageResults &&
               currentPageResults.map((vital, index) => {
                 return (
-                  <React.Fragment key={vital.id}>
+                  <React.Fragment key={index}>
```
Comment (FlorianRappl) [inline src/widgets/vitals/vitals-detailed-summary.component.tsx:98]: Usually having an unique `id` is better than the `index`. Are we sure there will not be swap operations, additions, or removals?

What's the motivation for the change (there are sometimes good reasons to pick an `index` instead of an `id`, maybe `id` does not even exist or is not unique - just wanted to understand)?

File: src/widgets/vitals/vitals-overview.component.tsx [hunk 5]
```diff
@@ -75,7 +75,7 @@ export default function VitalsOverview(props: VitalsOverviewProps) {
             <tbody>
               {currentVitals.map((vital, index) => {
                 return (
-                  <tr key={vital.id}>
+                  <tr key={index}>
```
Comment (FlorianRappl) [inline src/widgets/vitals/vitals-overview.component.tsx:78]: Same:-)
Curious, here, too!

**Review by FlorianRappl — COMMENTED** [review]
Looks great!

File: src/widgets/vitals/vitals-detailed-summary.component.tsx (hunk 4 above)
Comment (donaldkibet) [reply src/widgets/vitals/vitals-detailed-summary.component.tsx:98]: The id in this case is the encounter id which is not unique.

File: src/widgets/vitals/vitals-card.resource.tsx (hunk 3 above)
Comment (donaldkibet) [reply src/widgets/vitals/vitals-card.resource.tsx:128]: It is cast to Date but not instantiated, therefore at runtime it fails

Comment (FlorianRappl) [reply src/widgets/vitals/vitals-card.resource.tsx:128]: Well, but then the provided typing is wrong.

My recommendation would be to change it to whatever it is in reality, e.g., `string` or `number`. Otherwise, instantiate beforehand.

File: src/widgets/vitals/vital-record.component.tsx [hunk 6]
```diff
@@ -28,7 +30,16 @@ export default function VitalRecord(props: VitalRecordProps) {
   return (
     <>
       {vitalSigns && (
-        <SummaryCard name="Vital" styles={{ width: "100%" }}>
+        <SummaryCard
+          name="Vital"
```
Comment (FlorianRappl) [inline src/widgets/vitals/vital-record.component.tsx]: Can we translate `Vital`? Sorry I know it was not part of your PR - just having a dedicated eye on localization right now.

File: src/widgets/vitals/vital-record.component.tsx [hunk 7]
```diff
@@ -6,6 +6,8 @@ import { createErrorHandler } from "@openmrs/esm-error-handling";
 import styles from "./vital-record.css";
 import SummaryCard from "../../ui-components/cards/summary-card.component";
 import dayjs from "dayjs";
+import VitalsForm from "./vitals-form.component";
+import { openWorkspaceTab } from "../shared-utils";
 
 export default function VitalRecord(props: VitalRecordProps) {
   const [vitalSigns, setVitalSigns] = useState(null);
```
Comment (FlorianRappl) [inline src/widgets/vitals/vital-record.component.tsx]: What's the type here?

File: src/widgets/vitals/vital-record.component.tsx [hunk 8]
```diff
@@ -28,7 +30,16 @@ export default function VitalRecord(props: VitalRecordProps) {
   return (
     <>
       {vitalSigns && (
-        <SummaryCard name="Vital" styles={{ width: "100%" }}>
+        <SummaryCard
+          name="Vital"
+          styles={{ width: "100%" }}
+          editComponent={VitalsForm}
+          showComponent={() =>
+            openWorkspaceTab(VitalsForm, "Edit Vitals", {
```
Comment (FlorianRappl) [inline src/widgets/vitals/vital-record.component.tsx]: Translation needed

**Review by FlorianRappl — COMMENTED** [review]
I think we are close to finishing here.

Thanks @donaldkibet for the great work :rocket:!

File: src/widgets/vitals/vital-record.component.tsx (hunk 7 above)
Comment (donaldkibet) [reply src/widgets/vitals/vital-record.component.tsx]: `VitalsForm` is a functional component. `openWorkspaceTab` is function with return type `void`

File: src/widgets/heightandweight/heightandweight-record.component.tsx [hunk 9]
```diff
@@ -65,6 +65,15 @@ export default function HeightAndWeightRecord(
         <SummaryCard
           name={t("Height & Weight", "Height & Weight")}
           styles={{ width: "100%" }}
+          editComponent={VitalsForm}
+          showComponent={() =>
+            openWorkspaceTab(VitalsForm, "Edit Height and Weight", {
```
Comment (FlorianRappl) [inline src/widgets/heightandweight/heightandweight-record.component.tsx]: Can we introduce a translation for "Edit Height and Weight"?

**Review by FlorianRappl — COMMENTED** [review]
Approaching! :rocket:

File: src/widgets/vitals/vital-record.component.tsx (hunk 7 above)
Comment (FlorianRappl) [reply src/widgets/vitals/vital-record.component.tsx]: I was referring to `useState`. What type is `vitalSigns`?

(Sorry next time I'll be a bit more explicit, I thought the line number should be sufficient, but it can be ambiguous)

Comment (FlorianRappl) [reply src/widgets/vitals/vital-record.component.tsx]: Is this of type `PatientVitals`?

**Review by FlorianRappl — COMMENTED** [review]
Almost there - just this one typing enhancement (sorry, I know it was not introduced in this PR, so my intention here is just to improve the general codebase with these tiny things - thanks for your patience!).

**Review by FlorianRappl — APPROVED** [review]
Great work - thanks! :rocket:
